src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -212,11 +224,101 @@ public void close() throws IOException
               String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
           );
         }
+
+        if(!writerCurrentlyInUse)
+        {
+        	mergeWithSmoosher();
+        }
       }
     };
   }
+  
+  public OutputStream makeOutputStream(String filename) throws IOException
+  {
+    File retFile = createdFiles.get(filename);
+    if (retFile == null) {
+      retFile =  new File(baseDir, filename);
+      createdFiles.put(filename, retFile);
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } else  {
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } 
+  }
+  
+  public InputStream makeInputStream(String filename) throws IOException
+  {
+    final File retFile = createdFiles.get(filename);
+    return retFile == null ? null : new FileInputStream(retFile);
+  }
+  
+  private void mergeWithSmoosher() throws IOException
+  {
+	  //get processed elements from the stack and write.
+	 ArrayDeque<File> fileToProcess = files.clone();
+	 files = new ArrayDeque<>();
+	  while(! fileToProcess.isEmpty())
+	  {
+		  File file = fileToProcess.pop();
+		  add(file);
+		  file.delete();
+	  }
+  }
+  
+  private SmooshedWriter delegateSmooshedWriter(final String name,final long size) throws IOException
+  {	
+		return new SmooshedWriter () {
+			private int currOffset = 0;
+			private boolean open = true;
+			OutputStream writer= makeOutputStream(name);
+			
+			@Override
+			public void close() throws IOException
+			{
+				open = false;
+				writer.close();
+				files.add(createdFiles.get(name));
+				if(!writerCurrentlyInUse ) {
+				 mergeWithSmoosher();
+				}
+			}
+		    public int bytesLeft()
+		    {
+		      return (int) (size - currOffset);
+		    }
+
+		    @Override
+		    public int write(ByteBuffer buffer) throws IOException
+		    {
+		      WritableByteChannel channel = Channels.newChannel(writer);
+		      return addToOffset(channel.write(buffer));
+		    }
+
+		    @Override
+		    public int write(InputStream in) throws IOException
+		    {
+		      return addToOffset(Ints.checkedCast(ByteStreams.copy(in, writer)));
+		    }
+
+		    public int addToOffset(int numBytesWritten)
+		    {
+		      if (numBytesWritten > bytesLeft()) {
+		        throw new ISE("Wrote more bytes[%,d] than available[%,d]. Don't do that.", numBytesWritten, bytesLeft());
+		      }
+
+		      currOffset += numBytesWritten;
+
+		      return numBytesWritten;
+		    }
+
+			@Override
+			public boolean isOpen() {
+				return open;
+			}			
+		};
+	
+  }
 
-  @Override
+@Override
   public void close() throws IOException
   {
     Closeables.close(currOut, false);
```

Comment:
I think we should throw an exception in this method if there is a "temp" writer still open.